app/models/tag_set.rb

```diff
@@ -15,6 +15,20 @@ class TagSet < ApplicationRecord
   validates :name, presence: true, uniqueness: true
   validate :tag_group_adapter_types_must_match
 
+  scope :dual_index_tag_sets, -> { where.not(tag2_group: nil) }
+
+  scope :visible_dual_index_tag_sets,
+        -> do
+          dual_index_tag_sets.joins(:tag_group, :tag2_group).where(
+            tag_group: {
+              visible: true
+            },
+            tag2_group: {
+              visible: true
+            }
+          )
+        end
```

Comment:
These look good but I think you could simplify the naming to just `:dual_index` and `:visible_dual_index`.
(see examples of naming [here](https://www.rubyguides.com/2019/10/scopes-in-ruby-on-rails/) and [here](https://webcrunch.com/posts/how-to-use-scopes-ruby-on-rails))

Actually you can chain scopes so I wonder if it would be cleaner to just have `:dual_index` and `:visible`, and then your calling code could do `.dual_index.visible`. Don't know if there are performance implications - hopefully not!

Reply:
I would like to record a note here.

TagSet model **has one** tag_group and **has one** tag2_group. In order to support a unified access to dual and single index sets, it allows tag2_group to be null. TagSet has a method to check if itself is visible by checking the visibility of both tag_group and (if not null) tag2_group. The scoping here seems to be fetching the tag groups directly but I suppose having the tagset by name already would make tag_group and tag2_group available on the tag set.

Reply:
Thanks both, I have separated them for clarity, but I think `scope :visible_dual_index_tag_sets, -> { dual_index.visible }` is still needed as the scope is called from the `ranges.yml` file which doesn’t seem to handle chained scopes.